File: library/src/main/java/zhan/library/delegate/ScrollBodyDelegate.java
package zhan.library.delegate;

import android.view.MotionEvent;
import android.view.View;
import zhan.library.ScrollState;
import zhan.library.listener.OnStateChangeListener;



public class ScrollBodyDelegate extends ScrollerDelegate {

  private static final int NORMAL_TIME = 600;

  private int mMaxOffset;
  private float mLastY;
  private int mMoveY;

  private ScrollState mScrollState = ScrollState.SHOW;
  private View mTargetView;

  private OnStateChangeListener mOnStateChangeListener;

  public ScrollBodyDelegate(View targetView) {
    super(targetView);
    mTargetView = targetView;
  }

  public void setMaxOffset(int offset) {
    mMaxOffset = offset;
  }

  public void setState(ScrollState state) {
    mScrollState = state;
  }

  public ScrollState getState() {
    return mScrollState;
  }

  public boolean onTouchEvent(MotionEvent event) {
    if (mScrollState == ScrollState.HIDE) {
      return false;
    }
    float y = event.getY();
    switch (event.getAction()) {
      case MotionEvent.ACTION_DOWN:
        mLastY = y;
        break;

      case MotionEvent.ACTION_MOVE:
        int moveY = (int) (y - mLastY);
        if (mTargetView.getScrollY() <= 0 && moveY > 0) {
          int offset = moveY / 2;
          move(offset);
        }
        mLastY = y;
        break;

      case MotionEvent.ACTION_CANCEL:
      case MotionEvent.ACTION_UP:
        changeState();
        break;
    }
    return true;
  }

  private void move(int offset) {
    mScrollState = ScrollState.MOVE;
    if (mOnStateChangeListener != null) {
      mOnStateChangeListener.pullViewMove(mScrollState, -offset);
    }
    mTargetView.scrollBy(0, -offset);
  }

  private void hide() {
    mScrollState = ScrollState.HIDE;
    if (mOnStateChangeListener != null) {
      mOnStateChangeListener.pullViewHide(mScrollState);
    }
    mMoveY = mTargetView.getMeasuredHeight() + Math.abs(mTargetView.getScrollY());
    smoothScrollTo(0, mTargetView.getScrollY(), 0, -mMoveY, NORMAL_TIME * 3);
  }

  public void hide(int time) {
    mScrollState = ScrollState.HIDE;
    if (mOnStateChangeListener != null) {
      mOnStateChangeListener.pullViewHide(mScrollState);
    }
    mMoveY = mTargetView.getMeasuredHeight() + Math.abs(mTargetView.getScrollY());
    smoothScrollTo(0, mTargetView.getScrollY(), 0, -mMoveY, time);
  }

  private void show() {
    mScrollState = ScrollState.SHOW;
    if (mOnStateChangeListener != null) {
      mOnStateChangeListener.pullViewShow(mScrollState);
    }
    smoothScrollTo(0, mTargetView.getScrollY(), 0, -mTargetView.getScrollY(),
        mTargetView.getScrollY());
  }

  private void changeState() {
    if (Math.abs(mTargetView.getScrollY()) > mMaxOffset + 50) {
      hide();
    } else {
      show();
    }
  }

  public void open() {
    mScrollState = ScrollState.OPEN_START;
    if (mOnStateChangeListener != null) {
      mOnStateChangeListener.pullViewOpenStart();
    }
    smoothScrollTo(0, -mMoveY, 0, mMoveY, NORMAL_TIME);
    mTargetView.postDelayed(new Runnable() {
      @Override public void run() {
        mScrollState = ScrollState.OPEN_FINISH;
        if (mOnStateChangeListener != null) {
          mOnStateChangeListener.pullViewOpenFinish();
        }
      }
    }, NORMAL_TIME);
  }

  public void setOnStateChangeListener(OnStateChangeListener listener) {
    mOnStateChangeListener = listener;
  }
}

File: double-pull-delegate/src/main/java/zhan/pull/delegate/ScrollBodyDelegate.kt
package zhan.pull.delegate

import android.view.MotionEvent
import android.view.View
import zhan.pull.ScrollState
import zhan.pull.listener.OnStateChangeListener


class ScrollBodyDelegate(private val mTargetView: View) : ScrollerDelegate(mTargetView) {

    private var mMaxOffset: Int = 0
    private var mLastY: Float = 0.toFloat()
    private var mMoveY: Int = 0

    var state = ScrollState.SHOW

    private var mOnStateChangeListener: OnStateChangeListener? = null

    fun setMaxOffset(offset: Int) {
        mMaxOffset = offset
    }

    fun onTouchEvent(event: MotionEvent): Boolean {
        if (state === ScrollState.HIDE) {
            return false
        }
        val y = event.y
        when (event.action) {
            MotionEvent.ACTION_DOWN -> mLastY = y

            MotionEvent.ACTION_MOVE -> {
                val moveY = (y - mLastY).toInt()
                if (mTargetView.scrollY <= 0 && moveY > 0) {
                    val offset = moveY / 2
                    move(offset)
                }
                mLastY = y
            }

            MotionEvent.ACTION_CANCEL, MotionEvent.ACTION_UP -> changeState()
        }
        return true
    }

    private fun move(offset: Int) {
        state = ScrollState.MOVE
        if (mOnStateChangeListener != null) {
            mOnStateChangeListener!!.pullViewMove(state, -offset)
        }
        mTargetView.scrollBy(0, -offset)
    }

    private fun hide() {
        state = ScrollState.HIDE
        if (mOnStateChangeListener != null) {
            mOnStateChangeListener!!.pullViewHide(state)
        }
        mMoveY = mTargetView.measuredHeight + Math.abs(mTargetView.scrollY)
        smoothScrollTo(0, mTargetView.scrollY, 0, -mMoveY, NORMAL_TIME * 3)
    }

    fun hide(time: Int) {
        state = ScrollState.HIDE
        if (mOnStateChangeListener != null) {
            mOnStateChangeListener!!.pullViewHide(state)
        }
        mMoveY = mTargetView.measuredHeight + Math.abs(mTargetView.scrollY)
        smoothScrollTo(0, mTargetView.scrollY, 0, -mMoveY, time)
    }

    private fun show() {
        state = ScrollState.SHOW
        if (mOnStateChangeListener != null) {
            mOnStateChangeListener!!.pullViewShow(state)
        }
        smoothScrollTo(0, mTargetView.scrollY, 0, -mTargetView.scrollY,
                mTargetView.scrollY)
    }

    private fun changeState() {
        if (Math.abs(mTargetView.scrollY) > mMaxOffset + 50) {
            hide()
        } else {
            show()
        }
    }

    fun open() {
        state = ScrollState.OPEN_START
        if (mOnStateChangeListener != null) {
            mOnStateChangeListener!!.pullViewOpenStart()
        }
        smoothScrollTo(0, -mMoveY, 0, mMoveY, NORMAL_TIME)
        mTargetView.postDelayed({
            state = ScrollState.OPEN_FINISH
            if (mOnStateChangeListener != null) {
                mOnStateChangeListener!!.pullViewOpenFinish()
            }
        }, NORMAL_TIME.toLong())
    }

    fun setOnStateChangeListener(listener: OnStateChangeListener) {
        mOnStateChangeListener = listener
    }

    companion object {

        private val NORMAL_TIME = 600
    }
}
